compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs

```diff
@@ -135,18 +135,18 @@ impl TaitConstraintLocator<'_> {
             return;
         }
 
-        if let Some(hir_sig) = self.tcx.hir_node_by_def_id(item_def_id).fn_decl() {
-            if hir_sig.output.get_infer_ret_ty().is_some() {
-                let guar = self.tcx.dcx().span_delayed_bug(
-                    hir_sig.output.span(),
-                    "inferring return types and opaque types do not mix well",
-                );
-                self.found = Some(ty::OpaqueHiddenType {
-                    span: DUMMY_SP,
-                    ty: Ty::new_error(self.tcx, guar),
-                });
-                return;
-            }
+        // Function items with `_` in their return type already emit an error, skip any
+        // "non-defining use" errors for them.
+        if let Some(hir_sig) = self.tcx.hir_node_by_def_id(item_def_id).fn_sig()
```

Comment:
Can you change this to manually match on node kind, and only look for `Node::Item`/`Node::TraitItem`/`Node::ImplItem`/`Node::ForeignItem`?

The difference between when `Node::fn_sig` and `Node::fn_decl` return `Some(_)` is extremely subtle. It took me quite some time to even realize that's what changed in this PR.

Reply:
For `Node::ForeignItem` this isn't necessary because we already skip them here: https://github.com/rust-lang/rust/blob/30dfb9e046aeb878db04332c74de76e52fb7db10/compiler/rustc_hir_analysis/src/collect/type_of/opaque.rs#L262

I've added a test to ensure that is the case and we don't emit any extra errors for them.

Without `Node::ForeignItem`, the manual match is now just `Node::fn_sig` manually inlined and I don't think that's a lot better. The real problem here is the ambigous naming of `FnSig` vs `FnDecl`, where `FnSig` = `FnDecl` + `FnHeader` and `FnHeader` = effects (`const`/`aync`/`unsafe`) + call abi.

Reply:
What about renaming it to `fn_item_sig` or `fn_sig_of_item` or something? Something that distinguishes that it's grabbing just the signature of an item.

Reply:
For `fn_item_sig` I'd expect it to also include foreign function items, which `fn_sig` does not.

Reply:
I've updated the comment regarding `fn_sig` vs `fn_decl` and added a `debug_assert!` to hopefully make it more clear what's happening here.

And I'd rather not bikeshed a new name for `fn_sig` in this PR, so with that,
@rustbot review

For the record, here is the manual match version, which is literally just inlined `fn_sig`: https://github.com/rust-lang/rust/commit/79b8fa577579a4c7beeaa683a24326c1519d56e5

Reply:
I prefer the manual match version, but whatever